substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/sources/SourceCache.java

```diff
@@ -331,112 +310,64 @@ protected Path tryCacheFile(Path filePath, Class<?> clazz) {
-            moduleName = clazz.getModule().getName();
-        }
-
-        if (moduleName != null) {
-            for (String specialRootModule : specialRootModules) {
-                if (moduleName.equals(specialRootModule)) {
-                    // handle this module specially as it has intermediate dirs
-                    for (Path srcRoot : specialSrcRoots.get(specialRootModule)) {
-                        String srcRootGroup = srcRoot.subpath(1, 2).toString().replace(".", filePath.getFileSystem().getSeparator());
-                        if (filePath.toString().startsWith(srcRootGroup)) {
-                            Path sourcePath = extendPath(srcRoot, filePath);
-                            try {
-                                if (tryCheckCacheFile(sourcePath, targetPath)) {
-                                    return filePath;
-                                }
-                            } catch (IOException e) {
-                                /* delete the target file as it is invalid */
-                                targetPath.toFile().delete();
-                                /* have another go at caching it */
-                                return tryCacheFile(filePath, clazz);
-                            }
-                        }
-                    }
-                    break;
-                }
-            }
-        }
+    void maybeCacheSource(Source source, Path sourcePath, Path targetPath) {
+        // see if the target file already exists
+        File cachedFile = targetPath.toFile();
+        boolean doCopy;
 
-        for (SourceRoot root : srcRoots) {
-            final Path scopedFilePath;
-            if (moduleName != null && root.isJDK) {
-                scopedFilePath = Paths.get(moduleName, filePath.toString());
-            } else {
-                scopedFilePath = filePath;
-            }
-            final Path sourcePath = extendPath(root.path, scopedFilePath);
+        if (cachedFile.exists()) {
             try {
-                if (tryCheckCacheFile(sourcePath, targetPath)) {
-                    return filePath;
-                }
+                // only copy if the size is different or the target is older than the source
+                FileTime sourceTime = Files.getLastModifiedTime(sourcePath);
+                FileTime destTime = Files.getLastModifiedTime(targetPath);
+                long sourceSize = Files.size(sourcePath);
+                long destSize = Files.size(targetPath);
+                doCopy = sourceSize != destSize || destTime.compareTo(sourceTime) < 0;
             } catch (IOException e) {
-                /* delete the target file as it is invalid */
-                targetPath.toFile().delete();
-                /* have another go at caching it */
-                return tryCacheFile(filePath, clazz);
+                // copy anyway just in case
+                System.out.println("IOException " + e);
+                e.printStackTrace();
```

Comment:
I think this should be a warning.

Reply:
This was supposed to be removed. It was only there for debugging. We could possibly generate a (verbose level) log message here but I think it might be better to do that later where we catch errors when performing the copy.